File: telegram-service/src/main/java/de/nak/home_assistant/actions/DeviceAction.java
package de.nak.home_assistant.actions;

import com.pengrad.telegrambot.model.Message;
import com.pengrad.telegrambot.model.request.ForceReply;
import com.pengrad.telegrambot.request.BaseRequest;
import com.pengrad.telegrambot.request.EditMessageText;
import com.pengrad.telegrambot.request.PinChatMessage;
import com.pengrad.telegrambot.request.SendMessage;
import de.nak.home_assistant.CustomKeyboard;
import de.nak.home_assistant.models.service.ServiceRequest;
import de.nak.home_assistant.models.service.SimpleResponse;
import de.nak.home_assistant.models.telegram.CallbackData;
import de.nak.home_assistant.services.ServiceService;
import de.nak.home_assistant.services.database.DeviceService;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class DeviceAction {

    public static List<BaseRequest<?, ?>> generateMessages(long userId, long chatId, CallbackData cbData, Message inMessage, String data) {
        String deviceId = cbData.getDeviceId();
        int moduleId = cbData.getModuleId();

        List<BaseRequest<?, ?>> messages = new ArrayList<>();
        String function = cbData.getFunction();

        // Ask for data input if required and not given
        if (function.endsWith("*") && data == null) {
            SendMessage msg = new SendMessage(chatId, "Antworte auf diese Nachricht mit den Eingabe Daten")
                    .replyMarkup(new ForceReply());

            PinChatMessage pin = new PinChatMessage(chatId, inMessage.messageId()).disableNotification(true);

            if (inMessage.replyMarkup().inlineKeyboard()[0].length > 1) {
                EditMessageText edit = new EditMessageText(chatId, inMessage.messageId(), inMessage.text())
                        .replyMarkup(CustomKeyboard.generateFunctionButtons(Arrays.asList(cbData.getFunction()), deviceId, moduleId));

                messages.add(edit);
            }

            messages.add(pin);
            messages.add(msg);

            return messages;
        }

        DeviceService deviceService = new DeviceService();
        ServiceRequest serviceRequest = deviceService.getFunction(userId, moduleId, deviceId, cbData.getFunction());
        if (data != null) {
            serviceRequest.getBody().setAction(cbData.getFunction().replace("*", ""));
            serviceRequest.getBody().setData(data);
        }

        ServiceService service = new ServiceService(new SimpleResponse());
        SimpleResponse response = service.makeRequest(serviceRequest);

        String statusMessage = "";
        if (response.isSuccess()) {

            /* TODO: update edit message logic
            SendMessage sendMessage = DeviceSelect.generateMessage(userId, chatId, cbData);

            if (!sendMessage.getText().equals(inMessage.getText()) || sendMessage.getReplyMarkup() != inMessage.getReplyMarkup()) {

                messages.add(EditMessageText.builder()
                        .chatId(String.valueOf(chatId))
                        .messageId(inMessage.getMessageId())
                        .text(sendMessage.getText())
                        .replyMarkup((InlineKeyboardMarkup) sendMessage.getReplyMarkup())
                        .build()
                );
            }*/

            statusMessage = "Wir hatten erfolg!!";
        } else {
            statusMessage = "Hat irgendwie nicht geklappt!! " + response.getError();
        }

        messages.add(CustomKeyboard.generateMessageWithKeyboard(userId, chatId, statusMessage));
        return messages;
    }
}

File: telegram-service/src/actions/device-action.ts
import { Message } from "node-telegram-bot-api";
import bot from "../bot";
import { i18n } from "../i18n";
import { generateSendMessageOptions } from "../keyboard";
import { getDeviceFunction } from "../services/device";
import { makeServiceRequest } from "../services/service";
import { SimpleResponse } from "../services/typings";
import { CallbackData, getCallbackDataId } from "../telegram/callback-data";

export const sendDeviceAction = async (
  userId: number,
  chatId: number,
  cbData: CallbackData,
  { message_id }: Message,
  data?: string
) => {
  const translations = await i18n(userId);

  // Start input dialog
  if (cbData.data.endsWith("*") && !data) {
    await bot.sendMessage(chatId, translations.deviceAction.inputPrompt, {
      reply_markup: {
        force_reply: true,
      },
    });

    return bot.pinChatMessage(chatId, message_id.toString());
  }

  const { deviceId, moduleId } = getCallbackDataId(cbData);

  const serviceRequest = await getDeviceFunction(
    userId,
    moduleId,
    deviceId,
    cbData.data
  );

  if (!serviceRequest) {
    return bot.sendMessage(chatId, translations.deviceAction.databaseError);
  }

  const { body } = serviceRequest;
  if (data) {
    body.action = body.action.replace("*", "");
    body.data = data;
  }

  const res = await makeServiceRequest<SimpleResponse>(serviceRequest);

  const message = res.success
    ? translations.deviceAction.success
    : `${translations.deviceAction.error} ${res.error}`;

  return bot.sendMessage(
    chatId,
    message,
    await generateSendMessageOptions(userId)
  );
};
